### PR TITLE
Fix inset for large queries. Disable record level filtering

### PR DESCRIPTION
**Pull request by pwoody**

### CODE REVIEW

**Comment by pwoody** [comment]
@ash211 @robert3005

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 1]
```diff
@@ -30,6 +30,55 @@ import org.apache.spark.sql.types._
...
+    override def keep(value: T): Boolean = {
+      value != null && valueSet.contains(value)
+    }
+
+    // Drop when no value in the set is within the statistics range.
+    override def canDrop(statistics: Statistics[T]): Boolean = {
+      val statMax = statistics.getMax
+      val statMin = statistics.getMin
+      val statRange = com.google.common.collect.Range.closed(statMin, statMax)
+      !valueSet.exists(value => statRange.contains(value))
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: This scala function for set.exists returns fast on the first entry in the set that meets the condition?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 2]
```diff
@@ -394,6 +394,12 @@ class ParquetFileFormat
       SQLConf.PARQUET_INT96_AS_TIMESTAMP.key,
       int96AsTimestamp)
 
+
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: nit: extra newline here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 3]
```diff
@@ -30,6 +30,55 @@ import org.apache.spark.sql.types._
  */
 private[parquet] object ParquetFilters {
 
+  case class SetInFilter[T <: Comparable[T]](valueSet: Set[T])
+    extends UserDefinedPredicate[T] with Serializable {
+    private val min: T = valueSet.min
+    private val max: T = valueSet.max
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: are these used anywhere?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 4]
```diff
@@ -30,6 +30,55 @@ import org.apache.spark.sql.types._
...
+    // Drop when no value in the set is within the statistics range.
+    override def canDrop(statistics: Statistics[T]): Boolean = {
+      val statMax = statistics.getMax
+      val statMin = statistics.getMin
+      val statRange = com.google.common.collect.Range.closed(statMin, statMax)
+      !valueSet.exists(value => statRange.contains(value))
+    }
+
+    // Can only drop not(in(set)) when we are know that every element is in the set.
+    // From the statistics, we can only be assured of this when min == max.
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: editing nit: "when we are know".

is the "every element in the set" referring to "every element in the `valueSet` is in the statistics" ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 5]
```diff
@@ -30,6 +30,55 @@ import org.apache.spark.sql.types._
...
+    override def inverseCanDrop(statistics: Statistics[T]): Boolean = {
+      val statMax = statistics.getMax
+      val statMin = statistics.getMin
+      statMin == statMax && valueSet.contains(statMin)
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: I've got to admit I'm having a hard time following the inverses through this codepath.  This method is supposed to return true when the SetInFilter is inverted, like `not(in(set))`, and the statistics ensure that the values can never match the statistics?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 6]
```diff
@@ -30,6 +30,55 @@ import org.apache.spark.sql.types._
...
+      (n: String, v: Set[Any]) =>
+        FilterApi.userDefined(doubleColumn(n), SetInFilter(v.asInstanceOf[Set[java.lang.Double]]))
+    case StringType =>
+      (n: String, v: Set[Any]) =>
+        FilterApi.userDefined(binaryColumn(n),
+          SetInFilter(v.map(s => Binary.fromString(s.asInstanceOf[String]))))
+    case BinaryType =>
+      (n: String, v: Set[Any]) =>
+        FilterApi.userDefined(binaryColumn(n),
+          SetInFilter(v.map(e => Binary.fromReusedByteArray(e.asInstanceOf[Array[Byte]]))))
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: do the existing tests cover all 6 of these types for the in-set ?  I don't see new tests that do

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 7]
```diff
@@ -299,10 +348,7 @@ private[parquet] object ParquetFilters {
           .map(LogicalInverseRewriter.rewrite)
 
       case sources.In(name, values) if dataTypeOf.contains(name) =>
-        val eq = makeEq.lift(dataTypeOf(name))
-        values.flatMap { v =>
-          eq.map(_(name, v))
-        }.reduceLeftOption(FilterApi.or)
+        makeInSet.lift(dataTypeOf(name)).map(_(name, values.toSet))
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: nit: `.map` on separate line like above

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 8]
```diff
@@ -676,20 +673,35 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
...
 
+  test("Large In filters work with UDP") {
+    import testImplicits._
+    withSQLConf(ParquetOutputFormat.JOB_SUMMARY_LEVEL -> "ALL",
+      ParquetInputFormat.DICTIONARY_FILTERING_ENABLED -> "true") {
+      withTempPath { dir =>
+        val path = s"${dir.getCanonicalPath}/table1"
+        (1 to 1000).toDF().write.parquet(path)
+        val df = spark.read.parquet(path)
+        val filter = (1 to 500).map(i => i.toString).mkString(",")
```
Comment (ash211) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: make this 499 and use 499 and 501 below to ensure we're not accidentally negating anything

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 9]
```diff
@@ -30,6 +30,55 @@ import org.apache.spark.sql.types._
...
+
+    override def keep(value: T): Boolean = {
+      value != null && valueSet.contains(value)
+    }
+
+    // Drop when no value in the set is within the statistics range.
+    override def canDrop(statistics: Statistics[T]): Boolean = {
+      val statMax = statistics.getMax
+      val statMin = statistics.getMin
+      val statRange = com.google.common.collect.Range.closed(statMin, statMax)
```
Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: Scala has a range, let's use that

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala (hunk 5 above)
Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: Documentation has a great example https://github.com/palantir/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/filter2/predicate/UserDefinedPredicate.java#L57

**Pull request merged**